src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -78,22 +80,54 @@ internal class CompletionHandler : IRequestHandler<LSP.CompletionParams, LSP.Com
             var completionTrigger = await ProtocolConversions.LSPToRoslynCompletionTriggerAsync(request.Context, document, position, cancellationToken).ConfigureAwait(false);
 
             var list = await completionService.GetCompletionsAsync(document, position, completionTrigger, options: completionOptions, cancellationToken: cancellationToken).ConfigureAwait(false);
-            if (list == null)
+            if (list == null || list.Items.IsEmpty)
             {
                 return null;
             }
 
             var lspVSClientCapability = context.ClientCapabilities?.HasVisualStudioLspCapability() == true;
-
             var commitCharactersRuleCache = new Dictionary<ImmutableArray<CharacterSetModificationRule>, ImmutableArray<string>>();
 
             // Cache the completion list so we can avoid recomputation in the resolve handler
             var resultId = await _completionListCache.UpdateCacheAsync(list, cancellationToken).ConfigureAwait(false);
 
+            // Feature flag to enable the return of TextEdits instead of InsertTexts (will increase payload size).
+            // Flag is defined in VisualStudio\Core\Def\PackageRegistration.pkgdef.
+            // We also check against the CompletionOption for test purposes only.
+            Contract.ThrowIfNull(context.Solution);
+            var featureFlagService = context.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+            var returnTextEdits = featureFlagService.IsExperimentEnabled("Roslyn.LSP.Completion") ||
```

Comment:
the name should probably be in WellKnownExperimentNames http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Experiments/IExperimentationService.cs,2c4d693c654a75af